instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/filters/__init__.py

```diff
@@ -0,0 +1,218 @@
+# Copyright The OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+from typing import Callable, TypeVar, Union
+
+import grpc
+
+TCallDetails = TypeVar(
+    "TCallDetails", grpc.HandlerCallDetails, grpc.ClientCallDetails
+)
+Condition = Callable[[TCallDetails], bool]
+
+
+def _full_method(metadata):
+    name = ""
+    if isinstance(metadata, grpc.HandlerCallDetails):
+        name = metadata.method
+    # NOTE: replace here if there's better way to match cases to handle
+    # grpcext._interceptor._UnaryClientInfo/_StreamClientInfo
+    elif hasattr(metadata, "full_method"):
+        name = metadata.full_method
+    return name
+
+
+def _split_full_method(metadata):
+    name = _full_method(metadata)
+    service, method = os.path.split(name)
+    if service != "":
+        service = os.path.normpath(service)
+        service = service.lstrip("/")
+    return (service, method)
+
+
+def all_of(*args: Condition[TCallDetails]) -> Condition[TCallDetails]:
+    """Returns a filter function that returns True if all filter functions
+    assigned matches conditions.
+
+    Args:
+        args (function): a list of filter function
+
+    Returns:
+        A filter function that returns True if all filter functions
+    assigned matches conditions.
+    """
+
+    def filter_fn(metadata):
+        return all(func(metadata) for func in args)
+
+    return filter_fn
+
+
+def any_of(*args: Condition[TCallDetails]) -> Condition[TCallDetails]:
+    """Returns a filter function that returns True if any of filter functions
+    assigned matches conditions.
+
+    Args:
+        args (function): a list of filter function
+
+    Returns:
+        A filter function that returns True if any of filter functions
+    assigned matches conditions.
+    """
+
+    def filter_fn(metadata):
+        return any(func(metadata) for func in args)
+
+    return filter_fn
+
+
+def negate(func: Condition[TCallDetails]) -> Condition[TCallDetails]:
+    """Returns a filter function that negate the result of func
+
+    Args:
+        func (function): filter function to negate the result
+
+    Returns:
+        A filter function that negate the result of func
+    """
+
+    def filter_fn(metadata):
+        return not func(metadata)
+
+    return filter_fn
+
+
+def method_name(
+    name: Condition[Union[grpc.HandlerCallDetails, grpc.ClientCallDetails]]
```

Comment:
I may have made a bad comment before. But `name` here should be `str` I think?

Reply:
I also should have checked the type more carefully. Thank you for catching this.